# **Case Intro**

Term deposits are a major source of income for a bank. A term deposit is a cash investment held at a financial institution. Your money is invested for an agreed rate of interest over a fixed amount of time, or term. The bank has various outreach plans to sell term deposits to their customers such as email marketing, advertisements, telephonic marketing, and digital marketing.

Telephonic marketing campaigns still remain one of the most effective way to reach out to people. However, they require huge investment as large call centers are hired to actually execute these campaigns. Hence, it is crucial to identify the customers most likely to convert beforehand so that they can be specifically targeted via call.

The data is related to direct marketing campaigns (phone calls) of a Portuguese banking institution. The classification goal is to predict if the client will subscribe to a term deposit (variable y).

Content The data is related to the direct marketing campaigns of a Portuguese banking institution. The marketing campaigns were based on phone calls. Often, more than one contact to the same client was required, in order to access if the product (bank term deposit) would be ('yes') or not ('no') subscribed by the customer or not. The data folder contains two datasets:-

Bank.csv: 45,211 rows and 18 columns ordered by date (from May 2008 to November 2010)

Detailed Column Descriptions bank client data:

1 - age (numeric)

2 - job : type of job (categorical: "admin.","unknown","unemployed","management","housemaid","entrepreneur","student", "blue-collar","self-employed","retired","technician","services")

3 - marital : marital status (categorical: "married","divorced","single"; note: "divorced" means divorced or widowed)

4 - education (categorical: "unknown","secondary","primary","tertiary")

5 - default: has credit in default? (binary: "yes","no")

6 - balance: average yearly balance, in euros (numeric)

7 - housing: has housing loan? (binary: "yes","no")

8 - loan: has personal loan? (binary: "yes","no")

# **related with the last contact of the current campaign**:
9 - contact: contact communication type (categorical: "unknown","telephone","cellular") 10 - day: last contact day of the month (numeric)

11 - month: last contact month of year (categorical: "jan", "feb", "mar", …, "nov", "dec")

12 - duration: last contact duration, in seconds (numeric)

# **other attributes:**
13 - campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)

14 - pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric, -1 means client was not previously contacted)

15 - previous: number of contacts performed before this campaign and for this client (numeric)

16 - poutcome: outcome of the previous marketing campaign (categorical: "unknown","other","failure","success")

Output variable (desired target):

17 - y - has the client subscribed a term deposit? (binary: "yes","no")

Missing Attribute Values: None

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
df=pd.read_csv('https://raw.githubusercontent.com/ogut77/DataScience/main/data/Bank.csv',sep = ';')
df

age           job   marital  education default  balance housing loan  \
0       58    management   married   tertiary      no     2143     yes   no   
1       44    technician    single  secondary      no       29     yes   no   
2       33  entrepreneur   married  secondary      no        2     yes  yes   
3       47   blue-collar   married    unknown      no     1506     yes   no   
4       33       unknown    single    unknown      no        1      no   no   
...    ...           ...       ...        ...     ...      ...     ...  ...   
45206   51    technician   married   tertiary      no      825      no   no   
45207   71       retired  divorced    primary      no     1729      no   no   
45208   72       retired   married  secondary      no     5715      no   no   
45209   57   blue-collar   married  secondary      no      668      no   no   
45210   37  entrepreneur   married  secondary      no     2971      no   no   

         contact  day month  duration  campaign  pdays  previous poutcome    y  
0        unknown    5   may       261         1     -1         0  unknown   no  
1        unknown    5   may       151         1     -1         0  unknown   no  
2        unknown    5   may        76         1     -1         0  unknown   no  
3        unknown    5   may        92         1     -1         0  unknown   no  
4        unknown    5   may       198         1     -1         0  unknown   no  
...          ...  ...   ...       ...       ...    ...       ...      ...  ...  
45206   cellular   17   nov       977         3     -1         0  unknown  yes  
45207   cellular   17   nov       456         2     -1         0  unknown  yes  
45208   cellular   17   nov      1127         5    184         3  success  yes  
45209  telephone   17   nov       508         4     -1         0  unknown   no  
45210   cellular   17   nov       361         2    188        11    other   no  

[45211 rows x 17 columns]

In [ ]:
print(df.shape)
df.info()
df.isnull().sum()

(45211, 17)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45211 entries, 0 to 45210
Data columns (total 17 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   age        45211 non-null  int64 
 1   job        45211 non-null  object
 2   marital    45211 non-null  object
 3   education  45211 non-null  object
 4   default    45211 non-null  object
 5   balance    45211 non-null  int64 
 6   housing    45211 non-null  object
 7   loan       45211 non-null  object
 8   contact    45211 non-null  object
 9   day        45211 non-null  int64 
 10  month      45211 non-null  object
 11  duration   45211 non-null  int64 
 12  campaign   45211 non-null  int64 
 13  pdays      45211 non-null  int64 
 14  previous   45211 non-null  int64 
 15  poutcome   45211 non-null  object
 16  y          45211 non-null  object
dtypes: int64(7), object(10)
memory usage: 5.9+ MB


age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
y            0
dtype: int64

In [ ]:
#For object check the data 
for cn in df.columns:
  if(df[cn].dtype==object):
    print(df[cn].value_counts())

blue-collar      9732
management       9458
technician       7597
admin.           5171
services         4154
retired          2264
self-employed    1579
entrepreneur     1487
unemployed       1303
housemaid        1240
student           938
unknown           288
Name: job, dtype: int64
married     27214
single      12790
divorced     5207
Name: marital, dtype: int64
secondary    23202
tertiary     13301
primary       6851
unknown       1857
Name: education, dtype: int64
no     44396
yes      815
Name: default, dtype: int64
yes    25130
no     20081
Name: housing, dtype: int64
no     37967
yes     7244
Name: loan, dtype: int64
cellular     29285
unknown      13020
telephone     2906
Name: contact, dtype: int64
may    13766
jul     6895
aug     6247
jun     5341
nov     3970
apr     2932
feb     2649
jan     1403
oct      738
sep      579
mar      477
dec      214
Name: month, dtype: int64
unknown    36959
failure     4901
other       1840
success     1511
Name: poutcome, dtype: int64
n

In [ ]:
def Encoder(df):
          from sklearn import preprocessing
          columnsToEncode = list(df.select_dtypes(include=['category','object']))
          le = preprocessing.LabelEncoder()
          for feature in columnsToEncode:
              try:
                  df[feature] = le.fit_transform(df[feature])
              except:
                  print('Error encoding '+feature)
          return df

In [ ]:
df=Encoder(df)
df

age  job  marital  education  default  balance  housing  loan  contact  \
0       58    4        1          2        0     2143        1     0        2   
1       44    9        2          1        0       29        1     0        2   
2       33    2        1          1        0        2        1     1        2   
3       47    1        1          3        0     1506        1     0        2   
4       33   11        2          3        0        1        0     0        2   
...    ...  ...      ...        ...      ...      ...      ...   ...      ...   
45206   51    9        1          2        0      825        0     0        0   
45207   71    5        0          0        0     1729        0     0        0   
45208   72    5        1          1        0     5715        0     0        0   
45209   57    1        1          1        0      668        0     0        1   
45210   37    2        1          1        0     2971        0     0        0   

       day  month  duration  campaign  pdays  previous  poutcome  y  
0        5      8       261         1     -1         0         3  0  
1        5      8       151         1     -1         0         3  0  
2        5      8        76         1     -1         0         3  0  
3        5      8        92         1     -1         0         3  0  
4        5      8       198         1     -1         0         3  0  
...    ...    ...       ...       ...    ...       ...       ... ..  
45206   17      9       977         3     -1         0         3  1  
45207   17      9       456         2     -1         0         3  1  
45208   17      9      1127         5    184         3         2  1  
45209   17      9       508         4     -1         0         3  0  
45210   17      9       361         2    188        11         1  0  

[45211 rows x 17 columns]

In [ ]:
y = df['y'] #Output
X = df.drop('y',axis=1)
X

age  job  marital  education  default  balance  housing  loan  contact  \
0       58    4        1          2        0     2143        1     0        2   
1       44    9        2          1        0       29        1     0        2   
2       33    2        1          1        0        2        1     1        2   
3       47    1        1          3        0     1506        1     0        2   
4       33   11        2          3        0        1        0     0        2   
...    ...  ...      ...        ...      ...      ...      ...   ...      ...   
45206   51    9        1          2        0      825        0     0        0   
45207   71    5        0          0        0     1729        0     0        0   
45208   72    5        1          1        0     5715        0     0        0   
45209   57    1        1          1        0      668        0     0        1   
45210   37    2        1          1        0     2971        0     0        0   

       day  month  duration  campaign  pdays  previous  poutcome  
0        5      8       261         1     -1         0         3  
1        5      8       151         1     -1         0         3  
2        5      8        76         1     -1         0         3  
3        5      8        92         1     -1         0         3  
4        5      8       198         1     -1         0         3  
...    ...    ...       ...       ...    ...       ...       ...  
45206   17      9       977         3     -1         0         3  
45207   17      9       456         2     -1         0         3  
45208   17      9      1127         5    184         3         2  
45209   17      9       508         4     -1         0         3  
45210   17      9       361         2    188        11         1  

[45211 rows x 16 columns]

In [ ]:
from sklearn.model_selection import  train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=17)

In [ ]:
#Q1: Random Forest
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
forest = RandomForestClassifier(random_state=17 )
forest.fit(X_train, y_train)

#accuracy on test data
forest_pred = forest.predict(X_test)
print(accuracy_score(y_test, forest_pred) )

0.9025037600636999


In [ ]:
#Q1: XGBoost
from xgboost import XGBClassifier
model = XGBClassifier(random_state=17)
model.fit(X_train, y_train)

#accuracy on test data
xgb_pred = model.predict(X_test)
print(accuracy_score(y_test, xgb_pred))

0.90020348580023


In [ ]:
#Q1: Light GBM
import lightgbm as lgb
lgb_model = lgb.LGBMClassifier(random_state=17)
lgb_model.fit(X_train, y_train)

#accuracy on test data
print(accuracy_score(y_test, lgb_model.predict(X_test)))

0.9044501459789437


In [ ]:
#Q1: Gradient Boosting
from sklearn.ensemble import GradientBoostingClassifier
gbm_model = GradientBoostingClassifier(random_state=17)
gbm_model.fit(X_train, y_train)

#accuracy on test data
print(accuracy_score(y_test, gbm_model.predict(X_test)))

0.9018844554543042


Light GBM gives best accuracy compared to others. However the difference between them is very little.

In [ ]:
#Question 2
!pip install optuna

     |████████████████████████████████| 308 kB 4.6 MB/s 
     |████████████████████████████████| 210 kB 15.9 MB/s 
     |████████████████████████████████| 81 kB 8.6 MB/s 
     |████████████████████████████████| 78 kB 6.8 MB/s 
     |████████████████████████████████| 49 kB 5.7 MB/s 
     |████████████████████████████████| 113 kB 55.5 MB/s 
     |████████████████████████████████| 150 kB 49.5 MB/s 
  Created wheel for pyperclip: filename=pyperclip-1.8.2-py3-none-any.whl size=11137 sha256=910b8e340540a3c1b9a06fb3a7f33ad73b61abf17dae887f5eb63d053fe6442a
  Stored in directory: /root/.cache/pip/wheels/9f/18/84/8f69f8b08169c7bae2dde6bd7daf0c19fca8c8e500ee620a28
Successfully built pyperclip


In [ ]:
import optuna

import sklearn.datasets
import sklearn.ensemble
import sklearn.model_selection
import sklearn.svm

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=17)
# FYI: Objective functions can take additional arguments
# (https://optuna.readthedocs.io/en/stable/faq.html#objective-func-additional-args).
def objective(trial):
    x, y = X_train,y_train

    regressor_name = trial.suggest_categorical("regressor", ["GradientBoosting", "XGBoost","LightGBM","RandomForest"])
    if regressor_name == "GradientBoosting":
         max_depth = trial.suggest_int("max_depth",2,16)
         max_features = trial.suggest_int("max_features",2,16)
         regressor_obj = sklearn.ensemble.GradientBoostingRegressor(random_state=17,  max_depth=max_depth, max_features=max_features )
    
    elif regressor_name == "XGBoost":
         from xgboost import XGBRegressor
         max_depth = trial.suggest_int("max_depth",2,16)
         max_features = trial.suggest_int("max_features",2,16)
         regressor_obj = XGBRegressor(random_state=17,  max_depth=max_depth, max_features=max_features )
        
         

    elif regressor_name == "LightGBM":
         import lightgbm as lgb
         max_depth = trial.suggest_int("max_depth",2,16)
         max_features = trial.suggest_int("max_features",2,16)
         regressor_obj = lgb.LGBMRegressor(random_state=17,  max_depth=max_depth, max_features=max_features )
        
       
   
    else:
         max_depth = trial.suggest_int("max_depth",2,16)
         max_features = trial.suggest_int("max_features",2,16)
         regressor_obj = sklearn.ensemble.RandomForestRegressor(random_state=17,  max_depth=max_depth, max_features=max_features )

     

    r2=sklearn.model_selection.cross_val_score(regressor_obj, x, y, n_jobs=-1, cv=3).mean()
   
    return r2


if __name__ == "__main__":
    study = optuna.create_study(direction="maximize")
    study.optimize(objective, n_trials=100)
    print(study.best_trial)

[I 2022-03-28 10:35:40,906] A new study created in memory with name: no-name-46303fcd-cba2-4287-b373-1f3a80aad08a
[I 2022-03-28 10:35:46,373] Trial 0 finished with value: 0.34932171872922063 and parameters: {'regressor': 'GradientBoosting', 'max_depth': 3, 'max_features': 6}. Best is trial 0 with value: 0.34932171872922063.
[I 2022-03-28 10:35:47,725] Trial 1 finished with value: 0.38671654769148894 and parameters: {'regressor': 'LightGBM', 'max_depth': 7, 'max_features': 16}. Best is trial 1 with value: 0.38671654769148894.
[I 2022-03-28 10:36:02,715] Trial 2 finished with value: 0.35641397759461296 and parameters: {'regressor': 'RandomForest', 'max_depth': 14, 'max_features': 13}. Best is trial 1 with value: 0.38671654769148894.
[I 2022-03-28 10:36:03,794] Trial 3 finished with value: 0.38634580523343964 and parameters: {'regressor': 'LightGBM', 'max_depth': 13, 'max_features': 15}. Best is trial 1 with value: 0.38671654769148894.
[I 2022-03-28 10:36:08,207] Trial 4 finished with val

FrozenTrial(number=17, values=[0.3892534050390631], datetime_start=datetime.datetime(2022, 3, 28, 10, 37, 20, 666451), datetime_complete=datetime.datetime(2022, 3, 28, 10, 37, 21, 752700), params={'regressor': 'LightGBM', 'max_depth': 10, 'max_features': 9}, distributions={'regressor': CategoricalDistribution(choices=('GradientBoosting', 'XGBoost', 'LightGBM', 'RandomForest')), 'max_depth': IntUniformDistribution(high=16, low=2, step=1), 'max_features': IntUniformDistribution(high=16, low=2, step=1)}, user_attrs={}, system_attrs={}, intermediate_values={}, trial_id=17, state=TrialState.COMPLETE, value=None)


In [ ]:
#Q2: A
print(study.best_params)

{'regressor': 'LightGBM', 'max_depth': 10, 'max_features': 9}


In [ ]:
#Q2: B
import lightgbm as lgb
lgb_model = lgb.LGBMClassifier(random_state=17,max_depth=10,max_features=9)
lgb_model.fit(X_train, y_train)
# make predictions for test data
print(accuracy_score(y_test, lgb_model.predict(X_test)))

0.905511811023622


In [ ]:
#For Q3 and Q4 ,use the following data.
dr=pd.read_csv('https://raw.githubusercontent.com/ogut77/DataScience/main/data/diamond.csv')
dr

Carat Weight              Cut Color Clarity Polish Symmetry Report  \
0             1.10            Ideal     H     SI1     VG       EX    GIA   
1             0.83            Ideal     H     VS1     ID       ID   AGSL   
2             0.85            Ideal     H     SI1     EX       EX    GIA   
3             0.91            Ideal     E     SI1     VG       VG    GIA   
4             0.83            Ideal     G     SI1     EX       EX    GIA   
...            ...              ...   ...     ...    ...      ...    ...   
5995          1.03            Ideal     D     SI1     EX       EX    GIA   
5996          1.00        Very Good     D     SI1     VG       VG    GIA   
5997          1.02            Ideal     D     SI1     EX       EX    GIA   
5998          1.27  Signature-Ideal     G     VS1     EX       EX    GIA   
5999          2.19            Ideal     E     VS1     EX       EX    GIA   

      Price  
0      5169  
1      3470  
2      3183  
3      4370  
4      3171  
...     ...  
5995   6250  
5996   5328  
5997   6157  
5998  11206  
5999  30507  

[6000 rows x 8 columns]

In [ ]:
def Encoder(df):
          from sklearn import preprocessing
          columnsToEncode = list(df.select_dtypes(include=['category','object']))
          le = preprocessing.LabelEncoder()
          for feature in columnsToEncode:
              try:
                  df[feature] = le.fit_transform(df[feature])
              except:
                  print('Error encoding '+feature)
          return df

In [ ]:
dr=Encoder(dr)
dr

Carat Weight  Cut  Color  Clarity  Polish  Symmetry  Report  Price
0             1.10    2      4        2       3         0       1   5169
1             0.83    2      4        3       2         2       0   3470
2             0.85    2      4        2       0         0       1   3183
3             0.91    2      1        2       3         3       1   4370
4             0.83    2      3        2       0         0       1   3171
...            ...  ...    ...      ...     ...       ...     ...    ...
5995          1.03    2      0        2       0         0       1   6250
5996          1.00    4      0        2       3         3       1   5328
5997          1.02    2      0        2       0         0       1   6157
5998          1.27    3      3        3       0         0       1  11206
5999          2.19    2      1        3       0         0       1  30507

[6000 rows x 8 columns]

In [ ]:
y = dr['Price'] #Output
X = dr.drop('Price',axis=1)
X

Carat Weight  Cut  Color  Clarity  Polish  Symmetry  Report
0             1.10    2      4        2       3         0       1
1             0.83    2      4        3       2         2       0
2             0.85    2      4        2       0         0       1
3             0.91    2      1        2       3         3       1
4             0.83    2      3        2       0         0       1
...            ...  ...    ...      ...     ...       ...     ...
5995          1.03    2      0        2       0         0       1
5996          1.00    4      0        2       3         3       1
5997          1.02    2      0        2       0         0       1
5998          1.27    3      3        3       0         0       1
5999          2.19    2      1        3       0         0       1

[6000 rows x 7 columns]

In [ ]:
from sklearn.model_selection import  train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=17)

In [ ]:
#Question 3
def evalmetric(y,ypred):
 import numpy as np
 e = y - ypred 
 mse_f = np.mean(e**2)
 rmse_f = np.sqrt(mse_f)
 mae_f = np.mean(abs(e))
 mape_f = 100*np.mean(abs(e/y))
 r2_f = 1-(sum(e**2)/sum((y-np.mean(y))**2))
 print("MSE:", mse_f)
 print("RMSE:", rmse_f)
 print("MAE:",mae_f)
 print("MAPE:",mape_f)
 print("R-Squared:", r2_f)

In [ ]:
from sklearn.tree import DecisionTreeRegressor
dtr = DecisionTreeRegressor(random_state=17)
dtr.fit(X_train,y_train)
print('Test Data')
y_pred2 = dtr.predict(X_test)
evalmetric(y_test,y_pred2)

Test Data
MSE: 4222331.073863685
RMSE: 2054.83115458757
MAE: 1016.694373015873
MAPE: 7.833274374700561
R-Squared: 0.9592329752745201


In [ ]:
from sklearn.ensemble import RandomForestRegressor
rfr = RandomForestRegressor(random_state=17)
rfr.fit(X_train,y_train)
print('Test Data')
y_pred2 = rfr.predict(X_test)
evalmetric(y_test,y_pred2)

Test Data
MSE: 2074962.900565573
RMSE: 1440.4731516295515
MAE: 778.3143303628775
MAPE: 6.220426902140035
R-Squared: 0.9799660276771226


In [ ]:
from sklearn.ensemble import GradientBoostingRegressor
from time import time

gbr = GradientBoostingRegressor(random_state=17)
gbr.fit(X_train, y_train)
print('Test Data')
y_pred2 = gbr.predict(X_test)
evalmetric(y_test,y_pred2)

Test Data
MSE: 2702865.535065862
RMSE: 1644.0393958375396
MAE: 878.4993508475277
MAPE: 7.394144503096002
R-Squared: 0.9739035655494326


In [ ]:
from xgboost import XGBRegressor
xgr = XGBRegressor(random_state=17)
xgr.fit(X_train, y_train)
print('Test Data')
y_pred2 = xgr.predict(X_test)
evalmetric(y_test,y_pred2)

[10:53:08] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
Test Data
MSE: 2822838.991584785
RMSE: 1680.1306471774108
MAE: 881.3905965169271
MAPE: 7.3386749783130245
R-Squared: 0.9727452099437854


In [ ]:
from time import time
import lightgbm as lgb
t0 = time()
lgr = lgb.LGBMRegressor(random_state=17)
lgr.fit(X_train, y_train)
print('Test Data')
y_pred2 = lgr.predict(X_test)
evalmetric(y_test,y_pred2)

Test Data
MSE: 1940840.8294753155
RMSE: 1393.1406352107153
MAE: 701.0864858904304
MAPE: 5.433753614630323
R-Squared: 0.9812609895578275


In [ ]:
import sklearn.linear_model as skl_lm
regr = skl_lm.LinearRegression()
regr.fit(X_train,y_train)
print('Test Data')
y_pred2 = regr.predict(X_test)
evalmetric(y_test,y_pred2)

Test Data
MSE: 18292450.049325157
RMSE: 4276.967389322153
MAE: 2701.997276609754
MAPE: 30.68442425409791
R-Squared: 0.8233845829696962


Light GBM gives the best result with the R-squared = 0.9812609895578275

In [ ]:
#Question 4
import optuna

import sklearn.datasets
import sklearn.ensemble
import sklearn.model_selection
import sklearn.svm


# FYI: Objective functions can take additional arguments
# (https://optuna.readthedocs.io/en/stable/faq.html#objective-func-additional-args).
def objective(trial):
    x, y = X_train,y_train

    regressor_name = trial.suggest_categorical("regressor", ["GradientBoosting", "XGBoost","LightGBM","RandomForest"])
    if regressor_name == "GradientBoosting":
         max_depth = trial.suggest_int("max_depth", 2,7)
         max_features = trial.suggest_int("max_features", 2,7)
         regressor_obj = sklearn.ensemble.GradientBoostingRegressor(random_state=17,  max_depth=max_depth, max_features=max_features )
    
    elif regressor_name == "XGBoost":
         from xgboost import XGBRegressor
         max_depth = trial.suggest_int("max_depth", 2,7)
         max_features = trial.suggest_int("max_features", 2,7)
         regressor_obj = XGBRegressor(random_state=17,  max_depth=max_depth, max_features=max_features )
        
         

    elif regressor_name == "LightGBM":
         import lightgbm as lgb
         max_depth = trial.suggest_int("max_depth", 2,7)
         max_features = trial.suggest_int("max_features", 2,7)
         regressor_obj = lgb.LGBMRegressor(random_state=17,  max_depth=max_depth, max_features=max_features )
        
       
   
    else:
         max_depth = trial.suggest_int("max_depth", 2,7)
         max_features = trial.suggest_int("max_features", 2,7)
         regressor_obj = sklearn.ensemble.RandomForestRegressor(random_state=17,  max_depth=max_depth, max_features=max_features )

     

    r2=sklearn.model_selection.cross_val_score(regressor_obj, x, y, n_jobs=-1, cv=3).mean()
   
    return r2


if __name__ == "__main__":
    study = optuna.create_study(direction="maximize")
    study.optimize(objective, n_trials=100)
    print(study.best_trial)

[I 2022-03-28 10:57:27,047] A new study created in memory with name: no-name-6153bead-8258-4c3f-acba-c0ea899f1b90
[I 2022-03-28 10:57:29,489] Trial 0 finished with value: 0.9548589700649653 and parameters: {'regressor': 'XGBoost', 'max_depth': 2, 'max_features': 5}. Best is trial 0 with value: 0.9548589700649653.
[I 2022-03-28 10:57:30,596] Trial 1 finished with value: 0.873397050894793 and parameters: {'regressor': 'RandomForest', 'max_depth': 4, 'max_features': 3}. Best is trial 0 with value: 0.9548589700649653.
[I 2022-03-28 10:57:31,182] Trial 2 finished with value: 0.9842471016293839 and parameters: {'regressor': 'GradientBoosting', 'max_depth': 5, 'max_features': 3}. Best is trial 2 with value: 0.9842471016293839.
[I 2022-03-28 10:57:31,800] Trial 3 finished with value: 0.9758296442252078 and parameters: {'regressor': 'LightGBM', 'max_depth': 7, 'max_features': 3}. Best is trial 2 with value: 0.9842471016293839.
[I 2022-03-28 10:57:32,522] Trial 4 finished with value: 0.984430664

FrozenTrial(number=34, values=[0.9861873336374272], datetime_start=datetime.datetime(2022, 3, 28, 10, 57, 47, 76491), datetime_complete=datetime.datetime(2022, 3, 28, 10, 57, 47, 697038), params={'regressor': 'GradientBoosting', 'max_depth': 5, 'max_features': 6}, distributions={'regressor': CategoricalDistribution(choices=('GradientBoosting', 'XGBoost', 'LightGBM', 'RandomForest')), 'max_depth': IntUniformDistribution(high=7, low=2, step=1), 'max_features': IntUniformDistribution(high=7, low=2, step=1)}, user_attrs={}, system_attrs={}, intermediate_values={}, trial_id=34, state=TrialState.COMPLETE, value=None)


In [ ]:
#Q2: A
print(study.best_params)

{'regressor': 'GradientBoosting', 'max_depth': 5, 'max_features': 6}


In [ ]:
#Q2: B
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=17)
from sklearn.ensemble import GradientBoostingRegressor
from time import time

gbr = GradientBoostingRegressor(random_state=17,max_depth= 5,max_features= 6)
gbr.fit(X_train,y_train)
print('Test Data')
y_pred2 = gbr.predict(X_test)
evalmetric(y_test,y_pred2)

Test Data
MSE: 1542353.8550284477
RMSE: 1241.915397693598
MAE: 659.2526559755419
MAPE: 5.307389159446252
R-Squared: 0.985108420764872


R-squared is equal to 0.985108420764872